File: evolver/env.py
"""Minimal ``.env`` loader (no third-party dependency).

Reads ``KEY=VALUE`` lines from a ``.env`` file into ``os.environ`` before
:class:`~evolver.config.Config` is constructed (Config reads env in its field
defaults). Follows the usual dotenv conventions: blank lines and ``#`` comments
are ignored, an optional leading ``export`` is stripped, surrounding single/double
quotes are removed, and — by default — real environment variables already set
take precedence over the file (``override=False``).
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Dict, Optional, Union


def find_dotenv(start: Optional[Union[str, Path]] = None) -> Optional[Path]:
    """Search ``start`` (default cwd) and its parents for a ``.env`` file."""
    here = Path(start or Path.cwd()).resolve()
    for directory in (here, *here.parents):
        candidate = directory / ".env"
        if candidate.is_file():
            return candidate
    return None


def load_dotenv(
    path: Optional[Union[str, Path]] = None, override: bool = False
) -> Dict[str, str]:
    """Load a ``.env`` file into ``os.environ``; return the keys it set.

    If ``path`` is None the file is discovered by walking up from the current
    directory. Missing files are a no-op (returns ``{}``).
    """
    dotenv_path = Path(path) if path is not None else find_dotenv()
    if not dotenv_path or not dotenv_path.is_file():
        return {}

    loaded: Dict[str, str] = {}
    for raw in dotenv_path.read_text(encoding="utf-8").splitlines():
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        if line.startswith("export "):
            line = line[len("export "):].lstrip()
        if "=" not in line:
            continue
        key, _, value = line.partition("=")
        key = key.strip()
        value = value.strip()
        if len(value) >= 2 and value[0] == value[-1] and value[0] in ("'", '"'):
            value = value[1:-1]  # strip matching surrounding quotes (keep contents literal)
        if not key:
            continue
        if override or key not in os.environ:
            os.environ[key] = value
            loaded[key] = value
    return loaded

File: arbjs/src/env.js
/**
 * Minimal `.env` loader — the JS counterpart of `evolver/env.py`, and the reason
 * this package has no npm dependencies.
 *
 * Walks up from the current directory for a `.env`, reads `KEY=VALUE` lines into
 * `process.env`, and lets real environment variables win over the file.
 */

import { existsSync, readFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';

export function findDotenv(start = process.cwd()) {
    let dir = resolve(start);
    for (;;) {
        const candidate = join(dir, '.env');
        if (existsSync(candidate)) return candidate;
        const parent = dirname(dir);
        if (parent === dir) return null;
        dir = parent;
    }
}

export function loadDotenv({ path = null, override = false } = {}) {
    const file = path || findDotenv();
    if (!file || !existsSync(file)) return {};

    const loaded = {};
    for (const raw of readFileSync(file, 'utf8').split(/\r?\n/)) {
        let line = raw.trim();
        if (!line || line.startsWith('#')) continue;
        if (line.startsWith('export ')) line = line.slice('export '.length).trimStart();
        const eq = line.indexOf('=');
        if (eq < 0) continue;
        const key = line.slice(0, eq).trim();
        let value = line.slice(eq + 1).trim();
        if (value.length >= 2 && value[0] === value[value.length - 1] && (value[0] === "'" || value[0] === '"')) {
            value = value.slice(1, -1);
        }
        if (!key) continue;
        if (override || process.env[key] === undefined) {
            process.env[key] = value;
            loaded[key] = value;
        }
    }
    return loaded;
}
